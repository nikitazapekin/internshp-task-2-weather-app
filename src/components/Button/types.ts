export interface ButtonProps {
  text: string;
  handler: () => void;
  isActive?: boolean;
  isFullWidth?: boolean;
}
