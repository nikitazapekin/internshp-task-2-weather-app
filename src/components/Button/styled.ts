import { media } from "@styles/breakpoints";
import styled from "styled-components";

export const ButtonElement = styled.button<{ isActive?: boolean; isFullWidth?: boolean }>`
  cursor: pointer;
  padding: ${({ theme }) => theme.spaces.x}px ${({ theme }) => theme.spaces.lg}px;
  border: none;
  outline: none;
  border-radius: 10px;
  background-color: ${({ theme, isActive }) => (isActive ? theme.colors.blue : theme.colors.black)};
  color: ${({ theme }) => theme.colors.white};
  font-weight: ${({ theme }) => theme.fontWeights.regular};
  font-size: ${({ theme }) => theme.fontSizes.md};
  transition: 0.4s ease-in-out;
  min-width: 94px;
  width: ${({ isFullWidth }) => (isFullWidth ? "100%" : "auto")};

  &:hover {
    background-color: ${({ theme }) => theme.colors.blue};
  }

  ${media.lg`
   font-size: ${({ theme }) => theme.fontSizes.xxs};
   padding: ${({ theme }) => theme.spaces.xs}px ${({ theme }) => theme.spaces.xs}px;
  `}

  ${media.sm`
   font-size: ${({ theme }) => theme.fontSizes.xxs};
   padding: ${({ theme }) => theme.spaces.x}px ${({ theme }) => theme.spaces.x}px;
  `}
`;
