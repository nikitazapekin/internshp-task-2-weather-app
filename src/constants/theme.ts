export const theme = {
  container: 1187,
  deviceHeight: "100vh",
  colors: {
    black: "#000",
    blue: "#122e65",
    gray: "#D9D9D9",
    white: "#fff",
  },
  fontFamilies: {
    primary: "Inter",
    secondary: "sans-serif",
  },
  fontSizes: {
    xl: "41px",
    lg: "30px",
    md: "27px",
    sm: "22px",
    xs: "20px",
    xxs: "18px",
  },
  fontWeights: {
    regular: 400,
  },
  media: {
    xxl: 1400,
    xl: 1280,
    lg: 1024,
    md: 768,
    xh: 560,
    sm: 480,
  },
  zIndexes: {
    x: -1,
    xs: 0,
    sm: 1,
    md: 2,
  },
  spaces: {
    x: 5,
    xs: 10,
    sm: 15,
    md: 20,
    lg: 30,
    xl: 40,
    xxl: 50,
    g: 60,
    h: 70,
    xh: 80,
    eh: 90,
  },
  swiper: {
    item_gap: 5,
    margin_base: 2.5,
    vw_ratio: 0.5,
    scroll_snap_type: "x mandatory",
    scroll_snap_align: "start",
    flex_basis: "0 0",
    min_width: "0",
    container: 16.6,
  },
};
