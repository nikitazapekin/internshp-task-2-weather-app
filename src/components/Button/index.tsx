import { ButtonElement } from "./styled";
import type { ButtonProps } from "./types";

const Button = ({ text, handler, isActive, isFullWidth }: ButtonProps) => {
  return (
    <ButtonElement onClick={handler} $isActive={isActive} $isFullWidth={isFullWidth}>
      {text}
    </ButtonElement>
  );
};

export default Button;
