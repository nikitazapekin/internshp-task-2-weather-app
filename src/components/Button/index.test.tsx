import { theme } from "@constants/theme";
import { BUTTON_TEST } from "@mocks";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeProvider } from "styled-components";

import Button from ".";

const { DESCRIPTION, IT, CONSTANTS } = BUTTON_TEST;
const { RENDERS_CORRECT_TEXT, CALLS_HANDLER, APPLIES_FULL_WIDTH, MATCHES_SNAPSHOT } = IT;
const { DEFAULT_PROPS, STYLES } = CONSTANTS;

describe(`${DESCRIPTION}`, () => {
  test(`${RENDERS_CORRECT_TEXT}`, () => {
    render(
      <ThemeProvider theme={theme}>
        <Button {...DEFAULT_PROPS} />
      </ThemeProvider>
    );

    expect(screen.getByText(DEFAULT_PROPS.text)).toBeInTheDocument();
  });

  test(`${CALLS_HANDLER}`, async () => {
    const mockHandler = jest.fn();

    render(
      <ThemeProvider theme={theme}>
        <Button {...DEFAULT_PROPS} handler={mockHandler} />
      </ThemeProvider>
    );

    await userEvent.click(screen.getByRole("button"));
    expect(mockHandler).toHaveBeenCalledTimes(1);
  });

  test(`${APPLIES_FULL_WIDTH}`, () => {
    render(
      <ThemeProvider theme={theme}>
        <Button {...DEFAULT_PROPS} isFullWidth={true} />
      </ThemeProvider>
    );

    const button = screen.getByRole("button");

    expect(button).toHaveStyle(STYLES.FULL_WIDTH);
  });

  test(`${MATCHES_SNAPSHOT}`, () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Button {...DEFAULT_PROPS} />
      </ThemeProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
