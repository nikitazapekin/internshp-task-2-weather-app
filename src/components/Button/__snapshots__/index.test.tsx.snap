// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component matches snapshot 1`] = `
<button
  class="sc-bRKDuR bcDZqI"
>
  Test Button
</button>
`;
